### ISSUE TITLE
Check functions for integer overflow

### ISSUE THREAD

**Issue by gaborcsardi** [igraph/igraph#51]
In particular, all the functions that consider number of geodesics at some point might contain an integer overflow. Let's start with betweenness and edge_betweenness, I am sure that there are more, of course.

In case of integer overflow we just return an error with a new error code, IGRAPH_EINTOVERFLOW or something like that.

---

Imported from Launchpad using lp2gh.
- date created: 2009-08-03T18:27:24Z
- owner: gabor.csardi
- the launchpad url was https://bugs.launchpad.net/bugs/408521

**Comment by iosonofabio** [comment]
1. do we need a constant that stores the largest integer for this?
2. what type are the elements of `igraph_vector_t`? (I feel stupid asking this...)

**Comment by gaborcsardi** [comment]
1. There are constants in the standard C lib: INT_MAX, UINT_MAX, etc. But these are not very useful for checking integer overflow, because no int can be bigger than INT_MAX, by definition.
2. doubles

**Comment by iosonofabio** [comment]
Thank you. I thought one could do something like:

```C
int var_at_risk = ...;
int increment = ...;

if (increment > (INT_MAX - var_at_risk)) {
    IGRAPH_ERROR("Integer overflow", IGRAPH_EINTOVERFLOW);
}

var_at_risk += increment;
```

The number of geodesics in betweenness is an `igraph_vector_t`, so it's a double:

```C
    IGRAPH_VECTOR_INIT_FINALLY(&nrgeo, no_of_nodes);
```

so maybe I'm not understanding what kind of integer overflow you are afraid of?

**Comment by gaborcsardi** [comment]
Yeah, that would work I think. Maybe it is a double to avoid the overflow, and the issue is older?

**Comment by iosonofabio** [comment]
Alright, thanks, so I'll go around the codebase a little to identify integers that are at risk of overflowing

**Comment by gaborcsardi** [comment]
We can also deal with this if it comes up, I guess?

**Comment by iosonofabio** [comment]
I'll have a look around since I'm on it, then we can close this issue (marked for 0.9.0) and open new ones if that problem arises again.

**Comment by gaborcsardi** [comment]
Btw. another common way to check for integer overflow for an A + B sum is to check the sum is smaller than A. If it is, that's an overflow.

**Comment by iosonofabio** [comment]
Ok, found that `edge_betweenness` uses `unsigned long long int` which can in theory overflow. So I put checks for overflow there, in PR #1436

**Comment by szhorvat** [comment]
I made some progress on this:

 - Now we have safe integer operations macros: https://github.com/igraph/igraph/blob/develop/src/math/safe_intop.h
 - The core graph creation API, in `type_indexedgelist.c` should now be safe from overflow (and related crashes).
 - Overflow are still possible in most high-level graph creation functions. The typical mode of failure is an integer overflow when computing twice the number of edges, to be the size of an `edges` vector which is to be passed to `igraph_create()`. It seems that this problem needs to be fixed individually in each graph creation function.
     * [X] Deterministic graph creators (all fixed, or already immune to the problem)
     * [ ] Random graph generators (nothing is fixed so far).

**Comment by szhorvat** [comment]
The following stochastic generators have been verified to be overflow-safe:

 - `grg`
 - `tree_game_prufer`
 - `erdos_renyi_game_gnm`
 - `k_regular` calls `degree_sequence_game`, and needs no changes

The following need work:

 - [ ] `tree_game_loop_erased_random_walk`
 - [ ] ...

**Comment by GroteGnoom** [comment]
In igraph_i_barabasi_game_bag, `new_edges` is calculated by taking a `igraph_vector_int_sum`, which sums the degrees of vertices. Should we make an overflow-safe `igraph_vector_int_sum` ?

**Comment by szhorvat** [comment]
There is already an overflow-safe vector sum. Let's talk in the evening as I was also working on some of these functions.

**Comment by szhorvat** [comment]
I looked at it briefly, feel free to go ahead with this @GroteGnoom. We have `igraph_i_safe_vector_int_sum()` for sums. 

Preventing overflow is very tricky, and I'm sure we won't be able to make everything robust by 0.10. One problem I encountered recently was that some functions use reals instead of integers to prevent overflow on 32-bit, but at some point must convert the real back to integers (e.g. to get number of edges and reserve memory or to compute vertex IDs). This is yet another potential overflow. Then there's the issue of performance and complexity: safe operations are slow, and are difficult to write/read. The principle I try to follow is to minimize the overflow checks. It's best to do the smallest number of initial checks that will hopefully guarantee that no overflow can occur in the main part of the code. Sometimes this is easiest to achieve by making some sacrifices, e.g. restricting the vertex or edge count to some fraction of `IGRAPH_VCOUNT_MAX` or `IGRAPH_ECOUNT_MAX`. As an example, `IGRAPH_ECOUNT_MAX` itself is set not to `IGRAPH_INTEGER_MAX` but only half that, so that edge vectors (which are twice the length of the edge count) can be created safely. Thus with some functions it is even sufficient to just compute the edge count in an overflow safe manner and then check that it does not exceed `IGRAPH_ECOUNT_MAX`. In many (but not all) cases, there is no more chance for overflow after that.

**Comment by GroteGnoom** [comment]
If a user supplies a number of edges, should it be checked that its not greater than IGRAPH_ECOUNT_MAX?

**Comment by ntamas** [comment]
I guess so as its an O(1) check and we should probably have an `IGRAPH_CHECK_EDGE_COUNT()` macro anyway, but I'd leave it up to @szhorvat to decide.

**Comment by szhorvat** [comment]
Yes, it's theoretically nicer to fail fast, but for the vertex count it is not necessary. The max vertex count is currently `INTEGER_MAX - 1`. There is precisely one disallowed representable value. There is no harm in adding the check, but it's a lot of work (and extra code) for an infinitesimal gain.

This is a better argument for checking the edge count early, as the max edge count is `INTEGER_MAX/2`.

**Comment by szhorvat** [comment]
That said, we usually check that the vertex count is non-negative anyway, so we might as well wrap the edge and vertex count checks up into a macro that verifies that `0 <= n <= MAX`, just as Tamás suggested for edges.  Be careful though as a few functions don't allow a vertex count of `0`.
